File: pdfmaestro/annotation_overlay.py
"""
annotation_overlay.py — Transparent QWidget overlay on top of PDFViewer.
Handles mouse interaction for all annotation tools and emits committed signals.
Coordinate conversion between Qt screen space and PDF user space lives here.
"""
from __future__ import annotations

from PySide6.QtCore import Qt, Signal, QRectF, QPointF, QTimer
from PySide6.QtGui import QColor, QPainter, QPen, QBrush, QFont, QCursor
from PySide6.QtGui import QAction
from PySide6.QtWidgets import (
    QWidget, QInputDialog, QLineEdit, QMenu, QDialog,
    QVBoxLayout, QLabel, QDialogButtonBox, QComboBox,
)

# ── Tool constants ────────────────────────────────────────────────────────────
TOOL_POINTER   = "pointer"
TOOL_HIGHLIGHT = "highlight"
TOOL_NOTE      = "note"
TOOL_INK       = "ink"
TOOL_STAMP     = "stamp"
TOOL_REDACT    = "redact"

_CURSORS = {
    TOOL_POINTER:   Qt.ArrowCursor,
    TOOL_HIGHLIGHT: Qt.IBeamCursor,
    TOOL_NOTE:      Qt.PointingHandCursor,
    TOOL_INK:       Qt.CrossCursor,
    TOOL_STAMP:     Qt.PointingHandCursor,
    TOOL_REDACT:    Qt.CrossCursor,
}

_HIGHLIGHT_COLOR = QColor(255, 230, 0, 100)
_REDACT_COLOR    = QColor(0,   0,   0, 160)
_INK_COLOR       = QColor(0,   0,   200, 200)
_STAMP_COLOR     = QColor(0,   153, 0,  180)


class AnnotationOverlay(QWidget):
    """
    Transparent widget stacked on top of the viewer.
    Emits annotation_committed(...) with a dict describing what to write.
    MainWindow connects this signal to the pikepdf writers in annotations.py.
    """

    annotation_committed = Signal(dict)   # payload varies by tool

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setAttribute(Qt.WA_TransparentForMouseEvents, False)
        self.setAttribute(Qt.WA_NoSystemBackground, True)
        self.setAttribute(Qt.WA_TranslucentBackground, True)
        self.setMouseTracking(True)

        self._tool       = TOOL_POINTER
        self._page_idx   = 0
        self._page_rect  = QRectF()   # viewer coords: where the current page is drawn
        self._page_h_pt  = 792.0     # PDF page height in points (for Y-flip)
        self._zoom       = 1.0

        # Per-tool draw state
        self._drag_start: QPointF | None = None
        self._drag_cur:   QPointF | None = None
        self._ink_strokes: list[list[tuple[float, float]]] = []
        self._ink_current: list[tuple[float, float]] = []

    # ── Public API ────────────────────────────────────────────────────────────

    def set_tool(self, tool: str):
        self._tool = tool
        self.setCursor(QCursor(_CURSORS.get(tool, Qt.ArrowCursor)))
        # Pointer: pass events through to QGraphicsView for panning/scrolling.
        # Any other tool: overlay captures events for drawing.
        self.setAttribute(Qt.WA_TransparentForMouseEvents, tool == TOOL_POINTER)
        self._reset_draw_state()
        self.update()

    def set_page_context(self, page_idx: int, page_rect: QRectF,
                         page_h_pt: float, zoom: float):
        """Called by the viewer whenever the page layout changes."""
        self._page_idx  = page_idx
        self._page_rect = page_rect
        self._page_h_pt = page_h_pt
        self._zoom      = zoom

    # ── Coordinate helpers ────────────────────────────────────────────────────

    def _to_pdf(self, qpt: QPointF) -> tuple[float, float]:
        """Convert overlay widget coords → PDF user space (origin bottom-left)."""
        pr = self._page_rect
        if pr.width() == 0 or pr.height() == 0:
            return 0.0, 0.0
        rel_x = (qpt.x() - pr.x()) / pr.width()
        rel_y = (qpt.y() - pr.y()) / pr.height()
        # PDF Y-axis is bottom-up; Qt is top-down
        pdf_x = rel_x * self._page_w_pt()
        pdf_y = (1.0 - rel_y) * self._page_h_pt
        return pdf_x, pdf_y

    def _page_w_pt(self) -> float:
        pr = self._page_rect
        if pr.height() == 0:
            return 595.0
        return self._page_h_pt * pr.width() / pr.height()

    # ── Draw state ────────────────────────────────────────────────────────────

    def _reset_draw_state(self):
        self._drag_start  = None
        self._drag_cur    = None
        self._ink_strokes = []
        self._ink_current = []

    # ── Mouse events ─────────────────────────────────────────────────────────

    def mousePressEvent(self, event):
        try:
            if self._tool == TOOL_POINTER or not self._page_rect.contains(event.position()):
                event.ignore()
                return
            pos = event.position()
            if self._tool in (TOOL_HIGHLIGHT, TOOL_REDACT):
                self._drag_start = pos
                self._drag_cur   = pos
            elif self._tool == TOOL_INK:
                x, y = self._to_pdf(pos)
                self._ink_current = [(x, y)]
            elif self._tool == TOOL_NOTE:
                self._commit_note(pos)
            elif self._tool == TOOL_STAMP:
                self._commit_stamp(pos)
            self.update()
        except Exception:
            import traceback; traceback.print_exc()

    def mouseMoveEvent(self, event):
        try:
            pos = event.position()
            if self._tool in (TOOL_HIGHLIGHT, TOOL_REDACT) and self._drag_start:
                self._drag_cur = pos
                self.update()
            elif self._tool == TOOL_INK and self._ink_current:
                x, y = self._to_pdf(pos)
                self._ink_current.append((x, y))
                self.update()
        except Exception:
            import traceback; traceback.print_exc()

    def mouseReleaseEvent(self, event):
        try:
            pos = event.position()
            if self._tool == TOOL_HIGHLIGHT and self._drag_start:
                self._commit_highlight(self._drag_start, pos)
                self._reset_draw_state()
                self.update()
            elif self._tool == TOOL_REDACT and self._drag_start:
                self._commit_redact(self._drag_start, pos)
                self._reset_draw_state()
                self.update()
            elif self._tool == TOOL_INK:
                if len(self._ink_current) >= 2:
                    self._ink_strokes.append(list(self._ink_current))
                self._ink_current = []
                self.update()
        except Exception:
            import traceback; traceback.print_exc()

    def mouseDoubleClickEvent(self, event):
        try:
            if self._tool == TOOL_INK and self._ink_strokes:
                self._commit_ink()
                self._reset_draw_state()
                self.update()
        except Exception:
            import traceback; traceback.print_exc()

    # ── Commit helpers ────────────────────────────────────────────────────────

    def _commit_highlight(self, p1: QPointF, p2: QPointF):
        x0, y0 = self._to_pdf(QPointF(min(p1.x(), p2.x()), min(p1.y(), p2.y())))
        x1, y1 = self._to_pdf(QPointF(max(p1.x(), p2.x()), max(p1.y(), p2.y())))
        if abs(x1 - x0) < 2 or abs(y1 - y0) < 2:
            return
        # QuadPoints: four corners of the rectangle (BL, BR, TL, TR)
        quad = (x0, y0, x1, y0, x0, y1, x1, y1)
        self.annotation_committed.emit({
            "type":     "highlight",
            "page_idx": self._page_idx,
            "quads":    [quad],
            "color":    (1.0, 0.9, 0.0),
            "opacity":  0.4,
        })

    def _commit_note(self, pos: QPointF):
        x, y = self._to_pdf(pos)
        text, ok = QInputDialog.getMultiLineText(
            self, "Sticky Note", "Note text:"
        )
        if ok and text.strip():
            self.annotation_committed.emit({
                "type":     "note",
                "page_idx": self._page_idx,
                "x":        x,
                "y":        y,
                "contents": text.strip(),
            })

    def _commit_ink(self):
        if not self._ink_strokes:
            return
        self.annotation_committed.emit({
            "type":     "ink",
            "page_idx": self._page_idx,
            "strokes":  [list(s) for s in self._ink_strokes],
            "color":    (0.0, 0.0, 0.8),
            "width":    2.0,
        })

    def _commit_stamp(self, pos: QPointF):
        x, y = self._to_pdf(pos)
        dlg = _StampDialog(self)
        if dlg.exec():
            self.annotation_committed.emit({
                "type":     "stamp",
                "page_idx": self._page_idx,
                "x":        x,
                "y":        y - 40,
                "name":     dlg.selected_name,
            })

    def _commit_redact(self, p1: QPointF, p2: QPointF):
        x0, y0 = self._to_pdf(QPointF(min(p1.x(), p2.x()), min(p1.y(), p2.y())))
        x1, y1 = self._to_pdf(QPointF(max(p1.x(), p2.x()), max(p1.y(), p2.y())))
        if abs(x1 - x0) < 2 or abs(y1 - y0) < 2:
            return
        self.annotation_committed.emit({
            "type":     "redact",
            "page_idx": self._page_idx,
            "x0": x0, "y0": y0, "x1": x1, "y1": y1,
        })

    # ── Paint ─────────────────────────────────────────────────────────────────

    def paintEvent(self, _event):
        if self._tool == TOOL_POINTER:
            return
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)

        # Live drag rectangle for highlight / redact
        if self._tool in (TOOL_HIGHLIGHT, TOOL_REDACT) and self._drag_start and self._drag_cur:
            color = _HIGHLIGHT_COLOR if self._tool == TOOL_HIGHLIGHT else _REDACT_COLOR
            p.fillRect(QRectF(self._drag_start, self._drag_cur).normalized(), color)
            pen = QPen(color.darker(140), 1, Qt.DashLine)
            p.setPen(pen)
            p.drawRect(QRectF(self._drag_start, self._drag_cur).normalized())

        # Live ink strokes
        if self._tool == TOOL_INK:
            pen = QPen(_INK_COLOR, 2.5, Qt.SolidLine, Qt.RoundCap, Qt.RoundJoin)
            p.setPen(pen)
            for stroke in self._ink_strokes + ([self._pdf_to_screen_stroke(self._ink_current)]
                                                if self._ink_current else []):
                self._paint_screen_stroke(p, self._pdf_stroke_to_screen(stroke))
            if self._ink_strokes:
                p.setFont(QFont("sans-serif", 9))
                p.setPen(QPen(QColor(80, 80, 80)))
                p.drawText(self._page_rect.bottomLeft() + QPointF(4, -6),
                           "Double-click to commit ink")

    def _pdf_stroke_to_screen(self, stroke: list[tuple[float, float]]) -> list[QPointF]:
        pr = self._page_rect
        if pr.width() == 0 or pr.height() == 0:
            return []
        pw = self._page_w_pt()
        result = []
        for pdf_x, pdf_y in stroke:
            sx = pr.x() + (pdf_x / pw) * pr.width()
            sy = pr.y() + (1.0 - pdf_y / self._page_h_pt) * pr.height()
            result.append(QPointF(sx, sy))
        return result

    def _pdf_to_screen_stroke(self, pts_screen: list[tuple[float, float]]) -> list[tuple[float, float]]:
        # Already in PDF space — just return as-is for the combined call above
        return pts_screen

    def _paint_screen_stroke(self, p: QPainter, points: list[QPointF]):
        for i in range(len(points) - 1):
            p.drawLine(points[i], points[i + 1])


# ── Stamp picker dialog ───────────────────────────────────────────────────────

STAMP_NAMES = ["Approved", "Draft", "Confidential", "Final", "Void",
               "ForComment", "NotApproved", "Experimental"]


class _StampDialog(QDialog):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Insert Stamp")
        self.setFixedWidth(260)
        self.selected_name = "Approved"

        layout = QVBoxLayout(self)
        layout.addWidget(QLabel("Stamp type:"))
        self._combo = QComboBox()
        self._combo.addItems(STAMP_NAMES)
        layout.addWidget(self._combo)

        btns = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        btns.accepted.connect(self._on_ok)
        btns.rejected.connect(self.reject)
        layout.addWidget(btns)

    def _on_ok(self):
        self.selected_name = self._combo.currentText()
        self.accept()

File: src/annotationoverlay.cpp
#include "annotationoverlay.h"
#include <QMouseEvent>
#include <QKeyEvent>
#include <QPainter>
#include <QPen>
#include <QInputDialog>
#include <QComboBox>
#include <QDialogButtonBox>
#include <QDialog>
#include <QVBoxLayout>
#include <QLabel>
#include <QCursor>
#include <cmath>

static const QStringList STAMP_NAMES = {
    "Approved","Draft","Confidential","Final","Void","ForComment","NotApproved"
};

static QCursor cursorForTool(const QString& t) {
    if (t == "highlight")  return QCursor(Qt::IBeamCursor);
    if (t == "note")       return QCursor(Qt::PointingHandCursor);
    if (t == "ink")        return QCursor(Qt::CrossCursor);
    if (t == "stamp")      return QCursor(Qt::PointingHandCursor);
    if (t == "redact")     return QCursor(Qt::CrossCursor);
    if (t == "signature")  return QCursor(Qt::CrossCursor);
    if (t == "addtext")    return QCursor(Qt::IBeamCursor);
    return QCursor(Qt::ArrowCursor);
}

AnnotationOverlay::AnnotationOverlay(QWidget* parent)
    : QWidget(parent)
{
    setAttribute(Qt::WA_TransparentForMouseEvents, true);
    setAttribute(Qt::WA_NoSystemBackground, true);
    setAttribute(Qt::WA_TranslucentBackground, true);
    setMouseTracking(true);
}

void AnnotationOverlay::setSignatureImage(const QImage& img, double sigWPt) {
    m_sigImage = img;
    m_sigWPt   = sigWPt;
    m_sigHPt   = (img.height() > 0 && img.width() > 0)
                 ? sigWPt * double(img.height()) / double(img.width()) : 60.0;
    setFocusPolicy(Qt::StrongFocus);
    setFocus();
    update();
}

void AnnotationOverlay::setTool(const QString& tool) {
    m_tool = tool;
    setCursor(cursorForTool(tool));
    setAttribute(Qt::WA_TransparentForMouseEvents, tool == "pointer");
    resetDrawState();
    update();
}

void AnnotationOverlay::setPageContext(int pageIdx, const QRectF& pageRect,
                                        double pageHeightPt, double zoom) {
    m_pageIdx      = pageIdx;
    m_pageRect     = pageRect;
    m_pageHeightPt = pageHeightPt;
    m_zoom         = zoom;
}

double AnnotationOverlay::pageWidthPt() const {
    if (m_pageRect.height() == 0) return 595.0;
    return m_pageHeightPt * m_pageRect.width() / m_pageRect.height();
}

QPair<double,double> AnnotationOverlay::toPdf(const QPointF& pt) const {
    if (m_pageRect.width() == 0 || m_pageRect.height() == 0) return {0,0};
    double relX = (pt.x() - m_pageRect.x()) / m_pageRect.width();
    double relY = (pt.y() - m_pageRect.y()) / m_pageRect.height();
    return { relX * pageWidthPt(), (1.0 - relY) * m_pageHeightPt };
}

void AnnotationOverlay::resetDrawState() {
    m_dragging = false;
    m_inkStrokes.clear();
    m_inkCurrent.clear();
}

void AnnotationOverlay::mousePressEvent(QMouseEvent* event) {
    try {
        if (m_tool == "pointer") {
            event->ignore();
            return;
        }
        // Only reject clicks that are literally outside the visible widget area.
        // We do NOT guard on m_pageRect here: that rect is computed via
        // mapFromScene() and can be stale or have rounding discrepancies that
        // silently block every annotation click.  Coordinate clamping and
        // page-bounds validation happen in the QPDF layer instead.
        if (!rect().contains(event->position().toPoint())) {
            event->ignore();
            return;
        }
        QPointF pos = event->position();
        if (m_tool == "signature") {
            commitSignature(pos);
        } else if (m_tool == "addtext") {
            commitAddText(pos);
        } else if (m_tool == "highlight" || m_tool == "redact") {
            m_dragStart = pos;
            m_dragCur   = pos;
            m_dragging  = true;
        } else if (m_tool == "ink") {
            auto [px, py] = toPdf(pos);
            m_inkCurrent = { QPointF(px, py) };
        } else if (m_tool == "note") {
            commitNote(pos);
        } else if (m_tool == "stamp") {
            commitStamp(pos);
        }
        update();
    } catch (const std::exception& e) { qWarning() << e.what(); }
}

void AnnotationOverlay::mouseMoveEvent(QMouseEvent* event) {
    try {
        QPointF pos = event->position();
        if (m_tool == "signature") {
            m_sigGhostPos = pos;
            update();
        } else if ((m_tool == "highlight" || m_tool == "redact") && m_dragging) {
            m_dragCur = pos;
            update();
        } else if (m_tool == "ink" && !m_inkCurrent.isEmpty()) {
            auto [px, py] = toPdf(pos);
            m_inkCurrent.append(QPointF(px, py));
            update();
        }
    } catch (const std::exception& e) { qWarning() << e.what(); }
}

void AnnotationOverlay::mouseReleaseEvent(QMouseEvent* event) {
    try {
        QPointF pos = event->position();
        if (m_tool == "highlight" && m_dragging) {
            commitHighlight(m_dragStart, pos);
            m_dragging = false;
            update();
        } else if (m_tool == "redact" && m_dragging) {
            commitRedact(m_dragStart, pos);
            m_dragging = false;
            update();
        } else if (m_tool == "ink") {
            if (m_inkCurrent.size() >= 2)
                m_inkStrokes.append(m_inkCurrent);
            m_inkCurrent.clear();
            update();
        }
    } catch (const std::exception& e) { qWarning() << e.what(); }
}

void AnnotationOverlay::mouseDoubleClickEvent(QMouseEvent*) {
    try {
        if (m_tool == "ink" && !m_inkStrokes.isEmpty()) {
            commitInk();
            resetDrawState();
            update();
        }
    } catch (const std::exception& e) { qWarning() << e.what(); }
}

void AnnotationOverlay::commitHighlight(const QPointF& p1, const QPointF& p2) {
    // Pass raw viewport coordinates — MainWindow converts via viewportToPdf()
    // (mapToScene-based, always exact) instead of the stale m_pageRect-based toPdf().
    emit annotationCommitted({
        {"type", "highlight"},
        {"vpX0", p1.x()}, {"vpY0", p1.y()},
        {"vpX1", p2.x()}, {"vpY1", p2.y()}
    });
}

void AnnotationOverlay::commitNote(const QPointF& pos) {
    bool ok;
    QString text = QInputDialog::getMultiLineText(
        this, "Sticky Note", "Note text:", {}, &ok);
    if (!ok || text.trimmed().isEmpty()) return;
    auto [x, y] = toPdf(pos);
    emit annotationCommitted({
        {"type",     "note"},
        {"page",     m_pageIdx},
        {"x",        x}, {"y", y},
        {"contents", text.trimmed()}
    });
}

void AnnotationOverlay::commitInk() {
    if (m_inkStrokes.isEmpty()) return;
    QVariantList strokes;
    for (const auto& s : m_inkStrokes) {
        QVariantList pts;
        for (const auto& p : s) pts << QVariant::fromValue(p);
        strokes << QVariant(pts);
    }
    emit annotationCommitted({
        {"type",    "ink"},
        {"page",    m_pageIdx},
        {"strokes", strokes}
    });
}

void AnnotationOverlay::commitStamp(const QPointF& pos) {
    QDialog dlg(this);
    dlg.setWindowTitle("Insert Stamp");
    auto* layout = new QVBoxLayout(&dlg);
    layout->addWidget(new QLabel("Stamp type:"));
    auto* combo = new QComboBox;
    combo->addItems(STAMP_NAMES);
    layout->addWidget(combo);
    auto* btns = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(btns, &QDialogButtonBox::accepted, &dlg, &QDialog::accept);
    connect(btns, &QDialogButtonBox::rejected, &dlg, &QDialog::reject);
    layout->addWidget(btns);
    if (dlg.exec() != QDialog::Accepted) return;

    emit annotationCommitted({
        {"type",  "stamp"},
        {"vpX",   pos.x()},
        {"vpY",   pos.y()},
        {"w",     144.0}, {"h", 48.0},
        {"name",  combo->currentText()}
    });
}

void AnnotationOverlay::commitRedact(const QPointF& p1, const QPointF& p2) {
    emit annotationCommitted({
        {"type", "redact"},
        {"vpX0", p1.x()}, {"vpY0", p1.y()},
        {"vpX1", p2.x()}, {"vpY1", p2.y()}
    });
}

void AnnotationOverlay::keyPressEvent(QKeyEvent* event) {
    if (event->key() == Qt::Key_Escape && m_tool == "signature") {
        m_sigImage = QImage();
        setTool("pointer");
        emit signaturePlacementDone();
    } else {
        QWidget::keyPressEvent(event);
    }
}

void AnnotationOverlay::commitAddText(const QPointF& screenPos) {
    // Emit raw viewport coordinates — MainWindow converts via PdfViewer::viewportToPdf()
    // and runs the font-detection + dialog there (needs Poppler document access).
    emit annotationCommitted({
        {"type", "addtext"},
        {"vpX",  screenPos.x()},
        {"vpY",  screenPos.y()}
    });
}

void AnnotationOverlay::commitSignature(const QPointF& screenPos) {
    if (m_sigImage.isNull()) return;

    // Pass raw VIEWPORT coordinates to PdfViewer.
    // Do NOT convert through toPdf / m_pageRect here — that rect is the
    // overlay's cached rect of m_currentPage, which is wrong whenever the
    // page is partly scrolled off-screen or the user is near a page boundary.
    // PdfViewer uses mapToScene() (always exact) for the authoritative conversion.
    emit annotationCommitted({
        {"type",  "signature"},
        {"vpX",   screenPos.x()},   // viewport pixel X of ghost centre
        {"vpY",   screenPos.y()},   // viewport pixel Y of ghost centre
        {"sigW",  m_sigWPt},
        {"sigH",  m_sigHPt}
    });

    m_sigImage = QImage();
    setTool("pointer");
    emit signaturePlacementDone();
}

// ── Paint ─────────────────────────────────────────────────────────────────────

static QVector<QPointF> pdfStrokeToScreen(const QVector<QPointF>& stroke,
                                           const QRectF& pr, double pw, double ph) {
    QVector<QPointF> out;
    for (const auto& p : stroke) {
        double sx = pr.x() + (p.x() / pw) * pr.width();
        double sy = pr.y() + (1.0 - p.y() / ph) * pr.height();
        out.append({sx, sy});
    }
    return out;
}

void AnnotationOverlay::paintEvent(QPaintEvent*) {
    if (m_tool == "pointer") return;

    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    // Signature ghost — follows cursor
    if (m_tool == "signature" && !m_sigImage.isNull() && m_pageRect.isValid()) {
        double pxPerPt = (m_pageHeightPt > 0) ? m_pageRect.height() / m_pageHeightPt : 1.0;
        double ghostW  = m_sigWPt * pxPerPt;
        double ghostH  = m_sigHPt * pxPerPt;
        QRectF ghostRect(m_sigGhostPos.x() - ghostW / 2.0,
                         m_sigGhostPos.y() - ghostH / 2.0,
                         ghostW, ghostH);
        p.setOpacity(0.7);
        p.drawImage(ghostRect, m_sigImage);
        p.setOpacity(1.0);
        p.setPen(QPen(QColor(80, 120, 255), 1.5, Qt::DashLine));
        p.setBrush(Qt::NoBrush);
        p.drawRect(ghostRect);
        p.setFont(QFont("sans-serif", 9));
        p.setPen(QColor(60, 100, 220));
        p.drawText(ghostRect.bottomLeft() + QPointF(0.0, 18.0),
                   "Click to place  •  Esc to cancel");
        return;
    }

    if ((m_tool == "highlight" || m_tool == "redact") && m_dragging) {
        QColor col = (m_tool == "highlight")
                     ? QColor(255,230,0,100) : QColor(0,0,0,160);
        p.fillRect(QRectF(m_dragStart, m_dragCur).normalized(), col);
        p.setPen(QPen(col.darker(140), 1, Qt::DashLine));
        p.drawRect(QRectF(m_dragStart, m_dragCur).normalized());
    }

    if (m_tool == "ink") {
        p.setPen(QPen(QColor(0,0,200,200), 2.5, Qt::SolidLine,
                      Qt::RoundCap, Qt::RoundJoin));
        double pw = pageWidthPt();
        for (const auto& stroke : m_inkStrokes) {
            auto pts = pdfStrokeToScreen(stroke, m_pageRect, pw, m_pageHeightPt);
            for (int i = 1; i < pts.size(); ++i)
                p.drawLine(pts[i-1], pts[i]);
        }
        if (!m_inkCurrent.isEmpty()) {
            auto pts = pdfStrokeToScreen(m_inkCurrent, m_pageRect, pw, m_pageHeightPt);
            for (int i = 1; i < pts.size(); ++i)
                p.drawLine(pts[i-1], pts[i]);
        }
        if (!m_inkStrokes.isEmpty()) {
            p.setFont(QFont("sans-serif", 9));
            p.setPen(QColor(80,80,80));
            p.drawText(m_pageRect.bottomLeft() + QPointF(4,-6),
                       "Double-click to commit ink");
        }
    }
}
